### PR TITLE
Add support for CWE

### PR DESCRIPTION
**Pull request by ziadhany**
Reference: https://github.com/nexB/vulnerablecode/issues/651
Signed-off-by: Ziad <ziadhany2016@gmail.com>

### CODE REVIEW

**Comment by ziadhany** [comment]
Screenshot from 2022-06-27 20-19-05

File: vulnerabilities/models.py [hunk 1]
```diff
@@ -10,8 +10,9 @@
 import hashlib
 import json
 import logging
-import uuid
 
+from cwe import Database
+from django.contrib.postgres.fields import ArrayField
```
Comment (pombredanne) [inline vulnerabilities/models.py:15]: Do you mind to rather use a JSONField? this is more portable than the postgres-only ArrayField

**Review by pombredanne — CHANGES_REQUESTED** [review]
This would need to be rebased... also you may want to remove the UI part to treat it separately as the UI has changed extensively!
Also could we also get the CWE from the NVD importer? or is this something for later?
Finally two other considerations:
1. the CWE library maintainer has not much replied to data, so if we need this we would have to fok it
2. is there something else beyond CWE that would be about a more general concept of categories?

File: vulnerabilities/models.py (hunk 1 above)
Comment (ziadhany) [reply vulnerabilities/models.py:15]: No Problem, I will change it. but what is the best way to store CWE in the database as ( int or string ) 
{ "cwe_ids" : ['CWE-327', 'CWE-325' ]}
{ "cwe_ids" : [327 ,  325 ]}

**Comment by ziadhany** [comment]
> This would need to be rebased... also you may want to remove the UI part to treat it separately as the UI has changed extensively! Also could we also get the CWE from the NVD importer? or is this something for later? 

ok, no problem I will change this, but what about editing all importers not just NVD, I think I could handle this in a separate pull request.
>     1. the CWE library maintainer has not much replied to data, so if we need this we would have to fok it

oops, yes we can fork it. it isn't a complicated library, all data come from this [ database](https://github.com/Julian-Nash/cwe/blob/master/cwe/database.py), we need to make sure the database is updated so I think we should use this link 
https://cwe.mitre.org/data/downloads.html as our database. 

>     2. is there something else beyond CWE that would be about a more general concept of categories? 
I don't know but the main three categories are  
- **Software Development**
- Hardware Design
- Research Concepts

some External Mappings : 

- **CWE Top 25 (2022)**
- **OWASP Top Ten (2021)**
- **Software Fault Pattern Clusters**

**Comment by ziadhany** [comment]
Screenshot from 2022-09-11 01-38-04

**Comment by ziadhany** [comment]
Screenshot 2022-11-01 at 13-40-57 VulnerableCode Vulnerability Details - VCID-113g-6mxr-aaag

**Comment by pombredanne** [comment]
@ziadhany the cwe2 library is ready now: https://pypi.org/project/cwe2/

File: vulnerabilities/templates/vulnerability_details.html [hunk 2]
```diff
@@ -174,6 +173,36 @@
...
                         </table>
                     </div>
+
+                    <div class="has-text-weight-bold tab-nested-div ml-1 mb-1 mt-6">
+                        Weaknesses ({{ weaknesses|length }})
+                    </div>
+                     <div class="tab-nested-div">
+                        <table class="table is-bordered is-striped is-narrow is-hoverable is-fullwidth gray-header-border">
+                            <tr>
+                                <th> CWE id  </th>
```
Comment (pombredanne) [inline vulnerabilities/templates/vulnerability_details.html:183]: Use plain `CWE`

**Review by pombredanne — CHANGES_REQUESTED** [review]
Here are a few nitpickings for your review

File: vulnerabilities/improve_runner.py [hunk 3]
```diff
@@ -133,6 +134,10 @@ def process_inferences(inferences: List[Inference], advisory: Advisory, improver
                 fix=True,
             ).update_or_create()
 
+        if inference.weaknesses:
+            for cwe_id in inference.weaknesses:
+                Weakness.objects.update_or_create(cwe_id=cwe_id, vulnerabilities=vulnerability)
```
[the remark this answers is not in the document]
Comment (ziadhany) [reply vulnerabilities/improve_runner.py:139]: I think because we improve vulnerability one by one . Can you suggest any way to do this?

**Comment by pombredanne** [comment]
@ziadhany Could you rebase or merge on the latest main branch?

**Review by pombredanne — APPROVED** [review]
LGTM!

**Pull request merged**